test/particles/states.jl

```diff
@@ -48,21 +41,77 @@ test_broadcast(x::AbstractSpinOrPolarization) = x
-        @test QEDbase._as_svec(base_state(p(), d(), mom, SpinUp()))[1] ==
-            groundtruth_tuple[1]
-        @test QEDbase._as_svec(base_state(p(), d(), mom, SpinDown()))[1] ==
-            groundtruth_tuple[2]
+
+        @testset "tooling" begin
+            @test QEDbase._as_svec(base_state(p(), d(), mom, AllSpin())) isa SVector
+            @test QEDbase._as_svec(base_state(p(), d(), mom, SpinUp())) isa SVector
+            @test QEDbase._as_svec(base_state(p(), d(), mom, SpinDown())) isa SVector
+        end
+    end
+    @testset "spinor properties" begin
+        x, y, z = rand(RNG, 3)
+        m = mass(Electron())
+        P = SFourMomentum(sqrt(x^2 + y^2 + z^2 + m^2), x, y, z)
+
+        U = base_state(Electron(), Incoming(), P, AllSpin())
+        Ubar = base_state(Electron(), Outgoing(), P, AllSpin())
+        V = base_state(Positron(), Outgoing(), P, AllSpin())
+        Vbar = base_state(Positron(), Incoming(), P, AllSpin())
+
+        @testset "normatlisation" begin
+            for s1 in (1, 2)
+                for s2 in (1, 2)
+                    @test isapprox(Ubar[s1] * U[s2], 2 * m * (s1 == s2))
+                    @test isapprox(Vbar[s1] * V[s2], -2 * m * (s1 == s2))
+                    @test isapprox(Ubar[s1] * V[s2], 0.0)
+                    @test isapprox(Vbar[s1] * U[s2], 0.0)
+                end
+            end
+        end # normatlisation
+
+        @testset "completeness" begin
+            sumU = zero(DiracMatrix)
+            sumV = zero(DiracMatrix)
+            for spin in (1, 2)
+                sumU += U[spin] * Ubar[spin]
+                sumV += V[spin] * Vbar[spin]
+            end
+
+            @test isapprox(sumU, (slashed(P) + m * one(DiracMatrix)))
+            @test isapprox(sumV, (slashed(P) - m * one(DiracMatrix)))
+        end # completeness
+
+        @testset "diracs equation" begin
+            for spin in (1, 2)
+                @test isapprox(
+                    (slashed(P) - m * one(DiracMatrix)) * U[spin], zero(BiSpinor), atol=ATOL
+                )
+                @test isapprox(
+                    (slashed(P) + m * one(DiracMatrix)) * V[spin], zero(BiSpinor), atol=ATOL
+                )
+                @test isapprox(
+                    Ubar[spin] * (slashed(P) - m * one(DiracMatrix)),
+                    zero(AdjointBiSpinor),
+                    atol=ATOL,
+                )
+                @test isapprox(
+                    Vbar[spin] * (slashed(P) + m * one(DiracMatrix)),
+                    zero(AdjointBiSpinor),
+                    atol=ATOL,
+                )
+            end
+        end #diracs equation
+
+        @testset "sandwich" begin
+            for s1 in (1, 2)
+                for s2 in (1, 2)
+                    @test isapprox(
+                        SFourMomentum(Ubar[s1] * (GAMMA * U[s2])) * (s1 == s2),
+                        2 * P * (s1 == s2),
+                    )
```

Comment:
Is this trying to say that the two terms only need to be equal when `s1 == s2`?  If so, this seems like a somewhat hard-to-read way to do that. Maybe instead use `for (s1, s2) in ((1, 2), (2, 1))`?

Reply:
Thanks for pointing at this. Indeed, the line 109 is wrong, because this test should check if line 109 returns the same as 110 for `s1==s2` and zero otherwise. 

In 1a318dd I removed the `s1==s2` in line 109, which is now tested correctly.